src/Compilers/Core/Portable/InternalUtilities/SemaphoreExtensions.cs

```diff
@@ -0,0 +1,50 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Roslyn.Utilities
+{
+    internal static class SemaphoreExtensions
+    {
+        public static SemaphoreDisposer DisposableWait(this Semaphore semaphore, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (cancellationToken.CanBeCanceled)
+            {
+                int signalledIndex = WaitHandle.WaitAny(new[] { semaphore, cancellationToken.WaitHandle });
+                if (signalledIndex != 0)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+                    throw ExceptionUtilities.Unreachable;
+                }
+            }
+            else
+            {
+                semaphore.WaitOne();
+            }
+
+            return new SemaphoreDisposer(semaphore);
+        }
+
+        public static Task<SemaphoreDisposer> DisposableWaitAsync(this Semaphore semaphore, CancellationToken cancellationToken = default(CancellationToken))
```

Comment:
Does this need to exist? Pushing the blocking to the thread pool isn't nice to the thread pool (and only might slow things down since it presumes the thread pool is available to service a request.). I'd just delete this and use the non-async one at the use site.